DOI.js

```diff
@@ -93,52 +94,83 @@ function completeDOIs(doc) {
 	}
 }
 
-function retrieveDOIs(DOIs, doc) {
-	__num_DOIs = DOIs.length;
 
+var providers = [
+	"11645bd1-0420-45c1-badb-53fb41eeb753", //CrossRef
+	"9f1fb86b-92c8-4db7-b8ee-0b481d456428"  //DataCite
+];
+
+
+/*
+var providers = [
+	"9f1fb86b-92c8-4db7-b8ee-0b481d456428", //DataCite
+	"11645bd1-0420-45c1-badb-53fb41eeb753"  //CrossRef
+];
+*/
+
+function retrieveDOIs(DOIs, doc, provider) {
+	__num_DOIs = DOIs.length;
+	
 	for(var i=0, n=DOIs.length; i<n; i++) {
 		(function(doc, DOI) {
 			var translate = Zotero.loadTranslator("search");
-			translate.setTranslator("11645bd1-0420-45c1-badb-53fb41eeb753");
-	
+			translate.setTranslator(provider);
+			
 			var item = {"itemType":"journalArticle", "DOI":DOI};
 			translate.setSearch(item);
-	
+
 			// don't save when item is done
 			translate.setHandler("itemDone", function(translate, item) {
 				if (!item.title) {
-					Zotero.debug("No title available for " + DOI);
-					return;
+					Zotero.debug("No title available for " + item.DOI);
+					item.title = "[No Title]";
+					//return;
 				}
-				
-				item.repository = "CrossRef";
-				items[DOI] = item;
-				selectArray[DOI] = item.title;
+
+				items[item.DOI] = item;
+				selectArray[item.DOI] = item.title;
+				// done means not remaining anymore
+				Z.debug(item.DOI + "   " + remainingDOIs + "   " + provider + "    " + remainingDOIs.indexOf(item.DOI) );
+				remainingDOIs.splice(remainingDOIs.indexOf(item.DOI), 1);
+				Z.debug(remainingDOIs);
 			});
 	
 			translate.setHandler("done", function(translate) {
 				__num_DOIs--;
 				if(__num_DOIs <= 0) {
-					completeDOIs(doc);
+					Z.debug("done    " + providers.length);
```

Comment:
Yes, the debug messages are cryptic and may only helped me in debugging. (Here the debug message "debug 1" indicates that we are done with CrossRef and "debug 0" indicates that we are done with DataCite".)

In the last step I can delete all debug messages and adjust additionally added or deleted spaces. Let me know if you think that some debugging message should stay in the end. Here is an example for full debugging messages currently:
```
18:39:15 [
             "0": "10.1057/9780230391116"
             "1": "10.1073/pnas.1212247109"
             "2": "10.3205/mbi000337"
             "3": "10.3205/mbi000336"
             "4": "10.3205/mbi000354"
         ]
18:39:15 10.1057/9780230391116   10.1057/9780230391116,10.1073/pnas.1212247109,10.3205/mbi000337,10.3205/mbi000336,10.3205/mbi000354   11645bd1-0420-45c1-badb-53fb41eeb753    0
18:39:15 [
             "0": "10.1073/pnas.1212247109"
             "1": "10.3205/mbi000337"
             "2": "10.3205/mbi000336"
             "3": "10.3205/mbi000354"
         ]
18:39:15 10.1073/pnas.1212247109   10.1073/pnas.1212247109,10.3205/mbi000337,10.3205/mbi000336,10.3205/mbi000354   11645bd1-0420-45c1-badb-53fb41eeb753    0
18:39:15 [
             "0": "10.3205/mbi000337"
             "1": "10.3205/mbi000336"
             "2": "10.3205/mbi000354"
         ]
18:39:15 done    1
18:39:15 10.3205/mbi000336   10.3205/mbi000337,10.3205/mbi000336,10.3205/mbi000354   9f1fb86b-92c8-4db7-b8ee-0b481d456428    1
18:39:15 [
             "0": "10.3205/mbi000337"
             "1": "10.3205/mbi000354"
         ]
18:39:15 10.3205/mbi000354   10.3205/mbi000337,10.3205/mbi000354   9f1fb86b-92c8-4db7-b8ee-0b481d456428    1
18:39:15 [
             "0": "10.3205/mbi000337"
         ]
18:39:16 10.3205/mbi000337   10.3205/mbi000337   9f1fb86b-92c8-4db7-b8ee-0b481d456428    0
18:39:16 []
18:39:16 done    0
18:39:20 Translation successful
```